docs/syntax_and_semantics/case.md

```diff
@@ -146,3 +146,84 @@ when {String, Int32}
   # and the type of value2 is known to be an Int32
 end
 ```
+
+## Exhaustive case
+
+Using `in` instead of `when` produces an exhaustive case expression; in an exhaustive case, it is a compile-time error to omit any of the required `in` conditions. An exhaustive `case` cannot contain any `when` or `else` clauses.
+
+The compiler supports the following `in` conditions:
+
+### Union type checks
+
+If `case`'s expression is a union value, each of the union types may be used as a condition:
```

Comment:
Type-based expressions are not limited to union types but also work with type hierarchies.

Reply:
They don't work in virtual type hierarchies?